docs/torch/distrib.rst

```diff
@@ -0,0 +1,353 @@
+    ...             edge_label,
+    ...             feats[:, self.feature_dim :],
+    ...             nodes,
+    ...         )
+
+Experiment configuration
+========================
+Next step is to configure our experiment
+.. code-block:: python
+
+    >>> from deepgnn.graph_engine.snark.distributed import Server, Client as DistributedClient
+    >>> config = {
+    ...     "data_dir": "/tmp/ppi",  # graph.data_dir(),
+    ...     "device": torch.device("cpu"),
+    ...     "num_epochs": 2,
+    ...     "num_nodes": 56944,  # graph.NUM_NODES,
+    ...     "address": "localhost:9999",
+    ...     "feature_idx": 1,
+    ...     "feature_dim": 50,  # graph.FEATURE_DIM,
+    ...     "label_idx": 0,
+    ...     "label_dim": 1,
+    ...     "num_classes": 121,  # graph.NUM_CLASSES,
+    ...     "batch_size": 256,
+    ...     "fanout": [5, 5],
+    ... }
+    >>> s = Server(config["address"], config["data_dir"], 0, 1)
+    >>> def get_graph():
+    ...     return DistributedClient([config["address"]])
+    >>> config["get_graph"] = get_graph
+
+Train function
+==============
+We can now define our training function for a single epoch.
+
+.. code-block:: python
+
+    >>> from torch.nn.functional import binary_cross_entropy_with_logits
+    >>> def train_model(model, optimizer, loader):
+    ...     model.train()
+    ...     total_loss = 0
+    ...     for batch in loader:
+    ...         node_features, cols, rows, edge_label_index, edge_label = (
+    ...             batch[0][0],
+    ...             batch[2][0],
+    ...             batch[1][0],
+    ...             batch[3][0],
+    ...             batch[4][0],
+    ...         )
+    ...         edge_index = torch.stack([cols, rows], dim=0)
+    ...         optimizer.zero_grad()
+    ...         h = model(node_features, edge_index)
+    ...         h_src = h[edge_label_index[0]]
+    ...         h_dst = h[edge_label_index[1]]
+    ...         pred = (h_src * h_dst).sum(dim=-1)
+    ...         loss = binary_cross_entropy_with_logits(pred, edge_label)
+    ...         loss.backward()
+    ...         optimizer.step()
+    ...         total_loss += float(loss) * pred.size(0)
+    ...
+    ...     return total_loss
+
+Evaluation
+==========
+To evaluate model we train a sci-kit classifier on top of the embeddings on training predictions and then use that classifier on test dataset. It is charachterized with
+
+.. code-block:: python
+
+	>>> from sklearn.metrics import f1_score
+    >>> from sklearn.multioutput import MultiOutputClassifier
+    >>> from sklearn.linear_model import SGDClassifier
+    >>> def make_predictions(model, loader):
+    ...     ns, xs, ys = [], [], []
+    ...     for batch in loader:
+    ...         nodes, node_features, cols, rows, node_label = (
+    ...             batch[6][0],
+    ...             batch[0][0],
+    ...             batch[2][0],
+    ...             batch[1][0],
+    ...             batch[5][0],
+    ...         )
```

Comment:
Might be nicer like
node_features, rows, cols, _, _, node_label, nodes = batch[:, 0]

Reply:
Same for above

Reply:
Thanks!